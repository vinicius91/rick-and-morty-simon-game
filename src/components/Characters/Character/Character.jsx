// @flow

import * as React from 'react';
import { RegularCharacter, ActiveCharacter } from './Character.styles';

type Props = {
  id: number,
  active: boolean,
  imageUrl: string,
  color: string,
  colorDark: string,
  enabled: boolean,
  sound: string,
  characterClickHandler: (id: number) => void
};

const onClickHandler = (isEnabled, callback, id, audioRef) => {
  if (isEnabled) {
    callback(id);
    if (audioRef && audioRef.play) {
      audioRef.play();
    }
  }
};

const Character = (props: Props) => {
  let audioRef: ?HTMLAudioElement;
  const {
    characterClickHandler,
    active,
    imageUrl,
    color,
    colorDark,
    id,
    sound,
    enabled
  } = props;
  if (active) {
    return (
      <ActiveCharacter
        color={color}
        colorDark={colorDark}
        img={imageUrl}
        onClick={() => characterClickHandler(id)}
      >
        <audio src={sound} autoPlay />
      </ActiveCharacter>
    );
  }
  return (
    <RegularCharacter
      color={color}
      colorDark={colorDark}
      img={imageUrl}
      enabled={enabled}
      onClick={() =>
        onClickHandler(enabled, characterClickHandler, id, audioRef)
      }
    >
      <audio
        src={sound}
        ref={(ref: ?HTMLAudioElement) => {
          audioRef = ref;
        }}
      />
    </RegularCharacter>
  );
};

export default Character;
