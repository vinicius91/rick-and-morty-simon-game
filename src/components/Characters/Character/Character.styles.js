import styled, { css } from 'styled-components';

const clickMixin = css`
  :hover {
    transform: translate(2px, 2px);
  }

  :active {
    transform: translate(4px, 4px);
    box-shadow: ${props => `0 0 5px #fff, 
    inset 0 0 5px #fff, 
    0 0 20px ${props.color},
    inset 0 0 20px ${props.color}`};
  }
`;

export const RegularCharacter = styled.div`
  height: 20vh;
  flex: 0 1 40%;
  margin: 15px;
  background-size: cover;
  background-repeat: no-repeat;
  background-position: center center;
  border-radius: 10px;
  border: ${props => `3px solid var(--ui-bkgd, ${props.color})`};

  @media (min-width: 250px) and (max-width: 480px) {
    flex-basis: 40%;
    margin: 7px;
    width: 25vw;
    height: 35vw;
  }

  @media (min-width: 481px) and (max-width: 1600px) {
    flex-basis: 40%;
    height: 20vh;
  }

  background-image: ${props => `url(${props.img})`};
  box-shadow: ${props => `1px 0px ${props.colorDark}, 0px 1px ${props.color}, 2px 1px ${props.colorDark}, 1px 2px ${props.color},
    3px 2px ${props.colorDark}, 2px 3px ${props.color}, 4px 3px ${props.colorDark}, 3px 4px ${props.color},
    5px 4px ${props.colorDark}, 4px 5px ${props.color}, 6px 5px ${props.colorDark}, 5px 6px ${props.color}`};

  ${props => (props.enabled ? clickMixin : '')}
`;

export const ActiveCharacter = styled(RegularCharacter)`
  box-shadow: ${props => `0 0 5px #fff, 
    inset 0 0 5px #fff, 
    0 0 25px ${props.color},
    inset 0 0 25px ${props.color}`};

  :hover {
    transform: translate(0px, 0px);
  }
`;
