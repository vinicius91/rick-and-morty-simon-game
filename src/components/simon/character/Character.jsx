// @flow

import * as React from 'react';
import { RegularCharacter, ActiveCharacter } from './Character.styles';

type Props = {
  id: number,
  imageUrl: string,
  color: string,
  colorDark: string,
  active: boolean,
  characterClickHandler: (id: number) => void
};

const Character = (props: Props) => {
  const {
    characterClickHandler,
    active,
    imageUrl,
    color,
    colorDark,
    id
  } = props;
  if (active) {
    return (
      <ActiveCharacter
        color={color}
        colorDark={colorDark}
        img={imageUrl}
        onClick={() => characterClickHandler(id)}
      />
    );
  }
  return (
    <RegularCharacter
      color={color}
      colorDark={colorDark}
      img={imageUrl}
      onClick={() => characterClickHandler(id)}
    />
  );
};

export default Character;
