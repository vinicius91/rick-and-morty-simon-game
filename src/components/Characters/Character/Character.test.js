import React from 'react';
import renderer from 'react-test-renderer';
import Character from './Character';

describe('Character Component', () => {
  const fixture = {
    id: 1,
    imageUrl: '#',
    color: '#ffffff',
    colorDark: '#000000',
    sound: '#',
    active: false,
    enabled: true,
    characterClickHandler: () => {}
  };

  it('Should render default Character', () => {
    const component = renderer.create(<Character {...fixture} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render default Active Character', () => {
    const customFixture = { ...fixture, active: true };
    const component = renderer.create(<Character {...customFixture} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render default Disabled Character', () => {
    const customFixture = { ...fixture, enabled: false };
    const component = renderer.create(<Character {...customFixture} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
