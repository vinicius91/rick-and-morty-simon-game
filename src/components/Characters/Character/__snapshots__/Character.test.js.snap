// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character Component Should render default Active Character 1`] = `
<div
  className="Characterstyles__RegularCharacter-sc-77sfuh-0 Characterstyles__ActiveCharacter-sc-77sfuh-1 hLWHTt"
  color="#ffffff"
  onClick={[Function]}
>
  <audio
    autoPlay={true}
    src="#"
  />
</div>
`;

exports[`Character Component Should render default Character 1`] = `
<div
  className="Characterstyles__RegularCharacter-sc-77sfuh-0 bdvRsH"
  color="#ffffff"
  onClick={[Function]}
>
  <audio
    src="#"
  />
</div>
`;

exports[`Character Component Should render default Disabled Character 1`] = `
<div
  className="Characterstyles__RegularCharacter-sc-77sfuh-0 beBUvM"
  color="#ffffff"
  onClick={[Function]}
>
  <audio
    src="#"
  />
</div>
`;
